File: docs/canonicalize.py
"""
canonicalize.py — RFC 8785 subset for byte-stable JSON serialisation.

Companion to docs/audit_walker.py and docs/attest_verify.py: lets an
auditor in a Python-only environment re-verify a LocalFirst
attestation against the producer's canonical form without trusting
the producer's code.

Must stay logically identical to src/attest/canonicalize.js and the
inline copy in integrations/attest-view/viewer.js. The three
implementations exist so the schema is provably language-independent;
diverging them defeats the point.

Deviations from strict RFC 8785 (documented, intentional):
    - Floats use Python's json.dumps form. RFC 8785 mandates a
      specific ECMAScript Number.prototype.toString form which is
      equivalent for integers and most finite floats but not formally
      identical across all edge cases. The agent-attestation v1
      schema contains only integer counts and string hashes, so the
      distinction does not bite there.
    - Lone-surrogate handling matches Python json.dumps (escapes
      via \\uXXXX). JCS specifies the same.
"""

from __future__ import annotations

import json
from typing import Any


def canonicalize(value: Any) -> str:
    """Return the canonical-JSON string for ``value``.

    Rules:
      - object keys sorted lexicographically by UTF-16 code unit
        (Python's default ``sorted`` on strs is UTF-16-equivalent for
        the BMP, which covers every key in the v1 schema)
      - ``None`` ``True`` ``False`` map to ``null``/``true``/``false``
      - object members whose value is ``None`` are kept (they encode
        explicit nullable fields like ``policy.version``); members
        absent from the dict are not invented
      - arrays preserve order
      - rejects ``float('nan')``/``inf``, callables, types,
        non-string keys

    Raises ``ValueError`` on rejected inputs.
    """
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        if value != value or value in (float("inf"), float("-inf")):
            raise ValueError("canonicalize: non-finite number")
        # json.dumps for floats matches what V8's JSON.stringify
        # produces for finite floats in practice. For the v1 schema
        # this codepath is unused; kept for completeness.
        return json.dumps(value)
    if isinstance(value, str):
        # json.dumps emits a fully-escaped RFC 8259 string. Matches
        # what V8's JSON.stringify does for ASCII + most Unicode.
        return json.dumps(value, ensure_ascii=False)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(canonicalize(v) for v in value) + "]"
    if isinstance(value, dict):
        for k in value.keys():
            if not isinstance(k, str):
                raise ValueError(
                    f"canonicalize: non-string key {k!r}"
                )
        items = sorted(value.items(), key=lambda kv: kv[0])
        return "{" + ",".join(
            json.dumps(k, ensure_ascii=False) + ":" + canonicalize(v)
            for k, v in items
        ) + "}"
    raise ValueError(f"canonicalize: unsupported type {type(value).__name__}")

File: src/attest/canonicalize.js
'use strict';

/**
 * canonicalize.js — deterministic JSON serialisation for predicate/payload
 * byte-equivalence comparisons.
 *
 * This is a pragmatic subset of RFC 8785 (JSON Canonicalization Scheme):
 *   - object keys sorted lexicographically (UTF-16 code units; equivalent to
 *     default Array.prototype.sort on strings)
 *   - keys whose value is `undefined` are omitted (matches JSON.stringify)
 *   - arrays preserve order; `undefined` elements become `null`
 *   - numbers use V8's JSON.stringify form (RFC 8259 compliant for integers
 *     and finite decimals — our schema only uses integers, so this is exact)
 *   - strings use V8's JSON.stringify escaping
 *   - rejects: undefined at top level, non-finite numbers, functions, symbols
 *
 * Where this deviates from strict RFC 8785:
 *   - Float serialisation: RFC 8785 mandates a specific ECMAScript
 *     `Number.prototype.toString` form for non-integer floats. JSON.stringify
 *     in V8 matches this in practice but is not formally guaranteed. Our
 *     schema (agent-attestation v1) contains only integer counts and string
 *     hashes, so this is not an issue today. If we ever add float fields,
 *     swap this for a vetted JCS library.
 *   - Lone surrogate handling: JSON.stringify produces \uXXXX escapes which
 *     are valid RFC 8259. JCS specifies the same. Identical in practice.
 *
 * Why a tiny in-tree implementation instead of a dependency:
 *   - Must run unchanged in three places: Node (sign/verify), GitHub Action
 *     (no extra npm install before this runs), and the browser viewer
 *     (no build step). A copy-paste-ready ~30 lines is the cheapest path
 *     to "same bytes everywhere".
 *
 * IMPORTANT: keep this function logically identical to the inline copy in
 * integrations/attest-view/viewer.js. Any change here must be mirrored there.
 */

function canonicalize(value) {
  if (value === null) return 'null';
  switch (typeof value) {
    case 'boolean':
      return value ? 'true' : 'false';
    case 'number':
      if (!Number.isFinite(value)) {
        throw new Error('canonicalize: non-finite number');
      }
      return JSON.stringify(value);
    case 'string':
      return JSON.stringify(value);
    case 'object': {
      if (Array.isArray(value)) {
        return '[' + value.map(v =>
          canonicalize(v === undefined ? null : v)
        ).join(',') + ']';
      }
      const keys = Object.keys(value)
        .filter(k => value[k] !== undefined)
        .sort();
      return '{' + keys.map(k =>
        JSON.stringify(k) + ':' + canonicalize(value[k])
      ).join(',') + '}';
    }
    case 'undefined':
      throw new Error('canonicalize: undefined at top level');
    default:
      throw new Error('canonicalize: unsupported type ' + typeof value);
  }
}

module.exports = { canonicalize };
